client/asset/btc/btc.go

```diff
@@ -1417,24 +1622,60 @@ func (btc *ExchangeWallet) Confirmations(id dex.Bytes) (uint32, error) {
 // run pings for new blocks and runs the tipChange callback function when the
 // block changes.
 func (btc *ExchangeWallet) run(ctx context.Context) {
+	setBestBlock := func(hash string) *block {
+		latestBlock, err := btc.getBlockHeader(hash)
+		if err != nil {
+			btc.log.Debugf("unexpected getBlockHeader error for tip hash %q: %v", hash, err)
+			return nil
+		}
+		btc.tipMtx.Lock()
+		defer btc.tipMtx.Unlock()
+		btc.currentTip = &block{hash: latestBlock.Hash, height: latestBlock.Height}
+		return btc.currentTip
+	}
+
 	h, err := btc.node.GetBestBlockHash()
 	if err != nil {
 		btc.tipChange(fmt.Errorf("error initializing best block for %s: %v", btc.symbol, err))
 	}
-	tipHash := *h
+	setBestBlock(h.String())
+
 	ticker := time.NewTicker(blockTicker)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ticker.C:
-			h, err := btc.node.GetBestBlockHash()
+			newTipHash, err := btc.node.GetBestBlockHash()
 			if err != nil {
 				btc.tipChange(fmt.Errorf("failed to get best block hash from %s node", btc.symbol))
 				continue
 			}
-			if *h != tipHash {
-				tipHash = *h
-				btc.tipChange(nil)
+			btc.tipMtx.RLock()
+			prevTip := btc.currentTip
+			btc.tipMtx.RUnlock()
+			if newTipHash.String() == prevTip.hash {
+				continue
+			}
+			newTip := setBestBlock(newTipHash.String())
+			btc.tipChange(nil)
+			if newTip == nil {
+				continue // shouldn't happen but better to be safe
+			}
+			// Search for contract redemption in new blocks if there
+			// are contracts pending redemption.
+			btc.findRedemptionMtx.RLock()
+			pendingContractsCount := len(btc.findRedemptionQueue)
+			btc.findRedemptionMtx.RUnlock()
+			if pendingContractsCount == 0 {
+				continue
+			}
+			if newTip.height-prevTip.height > 1 {
+				// more than 1 block mined since last tip report,
+				// search all intermediate blocks for redemptions.
+				go btc.findRedemptionsInBlockRange(prevTip, newTip)
+			} else {
+				// just 1 new block, search lone block for redemptions.
+				go btc.findRedemptionsInBlockRange(newTip, newTip)
 			}
```

Comment:
I wouldn't think this should be necessary. Does just prev+1 to new work? What happens if reorg (I really don't know. This is one place I need to study a lot more.)?

Reply:
Yeah, this is an area I plan to re-visit, specifically better handling for re-orgs. If there's no re-org, prev+1 to new should work but I just feel better about rescanning the previous tip to ensure nothing is missed. Going with this formula, I should probably always just do prevTip to newTip (except when there's a re-org).